### PR TITLE
Gold Tests for successful examples

### PR DESCRIPTION
**Pull request by isovector**
This PR uses `tasty-golden` to generate gold tests for every passing example from `Examples.hs` (as marked in #25). It also pulls some of the testing utils out into their own `Miscellany` module, so that `Examples.hs` and `BasicTests.hs` can share them. The new tests are exposed in their own cabal entry, so it shouldn't affect any existing workflows.

It takes about 10 minutes on my machine to build this module. I suspect we could pull better compilation performance out of it if we separated it into several smaller modules, and compiled with `-j`, but that's a job for another PR.

Gold tests are written to `examples/test/gold/`, and can automatically regenerated via `stack test :gold-tests --test-arguments="--accept"` should they become out of date.

This PR doesn't merge yet; wanted to get some feedback on its mergeability before I go through the effort of merging.

### CODE REVIEW

**Comment by conal** [comment]
Thanks for this work! Some discussion points:

*   I ran several examples from Examples.hs yesterday, finding that many of those marked "failed" or "timeout" worked fine.
*   I wonder if a single module can serve both for somewhat-systematic testing (e.g., gold tests) as well as my ad hoc use for experimenting with transformations in the plugin (correctness and efficiency of generated computations). My impression of where you're going here is to *move* working examples from Examples.hs to BasicTest.hs. I may need to copy and re-tweak some of those tests back to Examples.hs, hopefully in way that's easy to do.
*   Perhaps we're better off separating example categories from testing and maybe manual, ad hoc testing from automated, systematic testing. Now, all three are in concat/examples.
*   If we're to factor automated testing into several modules, we may want test modules to correspond to categories or combinations of categories (e.g., AD + circuit). And maybe be able to select subsets of examples to compile per module, e.g., by having groups of examples with CPP conditionals. For instance, we could organize examples into groups numbered from zero upward, and there could be CPP flags for lower and upper bounds of which groups get compiled. Then various testing scripts could choose test modules and example ranges within those modules.

**Comment by isovector** [comment]
tl;dr: What would we do with `Examples` in an ideal world? What should we do with `Examples` today?


---

Curious about the failed tests. Some of them failed for missing imports, so I wonder if there's been an environmental change that fixed them. The timeout ones are likely just tests that were too slow for me -- I set a timeout of 30 seconds for them compile (some seem to take exponential time, which really adds up when you're compiling a couple hundred of these things). The good news is that the tooling in #25 should allow me to update the statuses of these tests quickly.

My plan here is indeed to move the tests out of `Examples`. I'm OK leaving them there as well, but there's going to be a cognitive burden in terms of which set of tests are canonical. We could also just say `Examples` is Conal's playground and leave it unmolested while working on systematizing the testing in other modules. IMO the real solution here is to figure out why the plugin doesn't work in GHCI and fix that, rather than doing REPL-like things with modules.

I agree strongly with your last two points, but would prefer to work towards them incrementally. The amount of work involved in sorting out `Examples` is already nigh overwhelming :)

**Comment by conal** [comment]
Thanks for the thoughts.

> IMO the real solution here is to figure out why the plugin doesn't work in GHCI and fix that, rather than doing REPL-like things with modules.

Amen to getting the plugin working with GHCi! It'd still be problematic to byte-compile a module that contains many examples if they're not mostly disabled, since the time-consuming compilation would happen when loading (byte-compiling) a module rather than performing a test. For Emacs, I guess I'd write some elisp code that zaps `#if`-hidden examples to the repl.

**Comment by conal** [comment]
@capn-freako Any thoughts on this PR and issues of testing?

**Comment by capn-freako** [comment]
Hi all,

Full disclosure: I don't really know much about proper software testing methodologies, at all. :(
(Hoping you guys can educate me.)

My experiences/hopes/frustrations/opinions are all largely in line w/ what I read above, specifically (and, so that you can check my understanding):

1. I would REALLY love to be able to run plug-in dependent code from the GHCI command prompt!!!

1. I agree with the need to maintain a testing "play" area that is allowed to live outside the rules we adopt to govern the process of changing our developing _canonical_ regression suite. The situation is still much too dynamic to get rid of that, now. That said, I'd love an alternative to the commenting/uncommenting of code we've been doing. That has always seemed fraught with peril to me.

1. I love the idea of extracting the more stable tests from the play area, above, to form our "golden", or "canonical", regression suite. I think having such a suite for, for instance, gating source code pushes will prove valuable, as more contributors join the effort.

Sorry I really can't do much other than reaffirm what you guys have already said.
If something new, or contrary, occurs to me, I'll post here so we can discuss.

Thanks for driving this forward, Sandy!

-db

**Comment by isovector** [comment]
If nobody has any concrete solutions, I'll suggest the following:

1) We merge this PR with the explicit goal of uncommenting as many lines of code as possible. Code that runs is >>>> code that can rot because it's a comment.
2) We continue pushing to remove all comments from `Examples`, but continue to use `Examples` as a place for interactively working on concat. Working on improving specific, old examples is done by yanking the relevant example from the tests.
3) Instead of commenting out new examples in `Examples`, they are instead moved into the gold testing framework.
4) We set up CI so that no human needs to actually run the sloooow tests by hand, and instead you'll get a notification on a PR if it breaks anything (or an email, if you have direct commit access).
5) I continue splitting out tests, sorting them by which categories they test (or by whatever criteria y'all would prefer).

If you are happy with (2) and (3), I don't mind continuing to fight with the tests and setting up CI. 

Thoughts?

**Comment by conal** [comment]
Thanks, for the comments, @capn-freako. I don't have much experience or insight about testing either, so I'm glad for the perspectives.

**Comment by conal** [comment]
Thanks for the articulation, @isovector! I'm happy with this plan. Please do continue, and I'll merge.

**Comment by isovector** [comment]
After syncing to ba63b7ea8dda55d8807815533cda001bdcdc305c, all of these tests seem to timeout or fail at runtime. This strikes me as a major regression -- is it something we'd expect?

The first error I get when trying to compile them:

```
ghc: panic! (the 'impossible' happened)
  (GHC version 8.0.2 for x86_64-unknown-linux):
        lam Case of boxer: bare unboxed var
  case exl   
         @ (->)
         @ Int
         @ Int
         ($p1BoolCat
            @ (->) ($p1EqCat @ (->) @ (Either () ()) $s$fEqCat(->)a_skw8))
         (f_sDo6
          `cast` ((R:OkSyn[0] ; Sym R:Ok(->)[0] <Int>_N,
                   R:OkSyn[0] ; Sym R:Ok(->)[0] <Int>_N)_R
                  :: ((Ok Syn Int, Ok Syn Int) :: Constraint)
                     ~R#
                     ((Ok (->) Int, Ok (->) Int) :: Constraint)))
         (exl
            @ (->)
            @ (Int, Int)
            @ Int
            ($p1BoolCat
               @ (->) ($p1EqCat @ (->) @ (Either () ()) $s$fEqCat(->)a_skw8))
            ($d(%,%)_XwBm
             `cast` ((Sym R:Ok(->)[0] <(Int, Int)>_N, Sym R:Ok(->)[0] <Int>_N)_R
                     :: ((Yes1 (Int, Int), Yes1 Int) :: Constraint)
                        ~R#
                        ((Ok (->) (Int, Int), Ok (->) Int) :: Constraint)))
            (exl
               @ (->)
               @ ((Int, Int), Int)
               @ Int
               ($p1BoolCat
                  @ (->) ($p1EqCat @ (->) @ (Either () ()) $s$fEqCat(->)a_skw8))
               (($fYes1ka @ * @ ((Int, Int), Int),
                 $p1&+&
                   @ *
                   @ Yes1
                   @ GenBuses
                   @ Int
                   (($p1(%,,%)
                       @ (Ok (Syn :**: (:>)) Int)
                       @ (Ok (Syn :**: (:>)) Int)
                       @ (Ok (Syn :**: (:>)) Bool)
                       $d(%,,%)_XrPz)
                    `cast` (Sub (D:R:Ok:**:[0] <(:>)>_N <Syn>_N)
                            ; (R:OkSyn[0] &+& D:R:Ok:>[0])_R <Int>_N
                            :: (Ok (Syn :**: (:>)) Int :: Constraint)
                               ~R#
                               ((&+&) Yes1 GenBuses Int :: Constraint))))
                `cast` ((Sym R:Ok(->)[0] <((Int, Int), Int)>_N,
                         Sym R:Ok(->)[0] <Int>_N)_R
                        :: ((Yes1 ((Int, Int), Int), Yes1 Int) :: Constraint)
                           ~R#
                           ((Ok (->) ((Int, Int), Int), Ok (->) Int) :: Constraint)))
               ww_ww1_ww2_ww3_Bh))
  of wild_ins8 { I# x1_insa ->
  case exr   
         @ (->)
         @ Int
         @ Int
         ($p1BoolCat
            @ (->) ($p1EqCat @ (->) @ (Either () ()) $s$fEqCat(->)a_skw8))
         (f_sDo6
          `cast` ((R:OkSyn[0] ; Sym R:Ok(->)[0] <Int>_N,
                   R:OkSyn[0] ; Sym R:Ok(->)[0] <Int>_N)_R
                  :: ((Ok Syn Int, Ok Syn Int) :: Constraint)
                     ~R#
                     ((Ok (->) Int, Ok (->) Int) :: Constraint)))
         (exl
            @ (->)
            @ (Int, Int)
            @ Int
            ($p1BoolCat
               @ (->) ($p1EqCat @ (->) @ (Either () ()) $s$fEqCat(->)a_skw8))
            ($d(%,%)_XwBm
             `cast` ((Sym R:Ok(->)[0] <(Int, Int)>_N, Sym R:Ok(->)[0] <Int>_N)_R
                     :: ((Yes1 (Int, Int), Yes1 Int) :: Constraint)
                        ~R#
                        ((Ok (->) (Int, Int), Ok (->) Int) :: Constraint)))
            (exl
               @ (->)
               @ ((Int, Int), Int)
               @ Int
               ($p1BoolCat
                  @ (->) ($p1EqCat @ (->) @ (Either () ()) $s$fEqCat(->)a_skw8))
               (($fYes1ka @ * @ ((Int, Int), Int),
                 $p1&+&
                   @ *
                   @ Yes1
                   @ GenBuses
                   @ Int
                   (($p1(%,,%)
                       @ (Ok (Syn :**: (:>)) Int)
                       @ (Ok (Syn :**: (:>)) Int)
                       @ (Ok (Syn :**: (:>)) Bool)
                       $d(%,,%)_XrPz)
                    `cast` (Sub (D:R:Ok:**:[0] <(:>)>_N <Syn>_N)
                            ; (R:OkSyn[0] &+& D:R:Ok:>[0])_R <Int>_N
                            :: (Ok (Syn :**: (:>)) Int :: Constraint)
                               ~R#
                               ((&+&) Yes1 GenBuses Int :: Constraint))))
                `cast` ((Sym R:Ok(->)[0] <((Int, Int), Int)>_N,
                         Sym R:Ok(->)[0] <Int>_N)_R
                        :: ((Yes1 ((Int, Int), Int), Yes1 Int) :: Constraint)
                           ~R#
                           ((Ok (->) ((Int, Int), Int), Ok (->) Int) :: Constraint)))
               ww_ww1_ww2_ww3_Bh))
  of wild1_insc { I# y1_inse ->
  let {      
    $j_s4ICn :: Int# -> Int -> Int
    [LclId,  
     Arity=2,
     Str=DmdType,
     Unf=Unf{Src=<vanilla>, TopLvl=False, Value=True, ConLike=True,
             WorkFree=True, Expandable=True, Guidance=IF_ARGS [0 0] 690 0}]
    $j_s4ICn =
      \ (x1_Xo48 [OS=OneShot] :: Int#)
        (wild_Xo45 [OS=OneShot]
           :: Int
           Unf=Unf{Src=InlineStable, TopLvl=False, Value=True, ConLike=True,
                   WorkFree=True, Expandable=True,
                   Guidance=ALWAYS_IF(arity=0,unsat_ok=True,boring_ok=False)
                   Tmpl= I# x1_Xo48}) ->
        case exr
               @ (->)
               @ (Int, Int)
               @ Int
               ($p1BoolCat
                  @ (->) ($p1EqCat @ (->) @ (Either () ()) $s$fEqCat(->)a_skw8))
               ($d(%,%)_XwBm
                `cast` ((Sym R:Ok(->)[0] <(Int, Int)>_N, Sym R:Ok(->)[0] <Int>_N)_R
                        :: ((Yes1 (Int, Int), Yes1 Int) :: Constraint)
                           ~R#
                           ((Ok (->) (Int, Int), Ok (->) Int) :: Constraint)))
               (exl
                  @ (->)
                  @ ((Int, Int), Int)
                  @ Int
                  ($p1BoolCat
                     @ (->) ($p1EqCat @ (->) @ (Either () ()) $s$fEqCat(->)a_skw8))
                  (($fYes1ka @ * @ ((Int, Int), Int),
                    $p1&+&
                      @ *
                      @ Yes1
                      @ GenBuses
                      @ Int
                      (($p1(%,,%)
                          @ (Ok (Syn :**: (:>)) Int)
                          @ (Ok (Syn :**: (:>)) Int)
                          @ (Ok (Syn :**: (:>)) Bool)
                          $d(%,,%)_XrPz)
                       `cast` (Sub (D:R:Ok:**:[0] <(:>)>_N <Syn>_N)
                               ; (R:OkSyn[0] &+& D:R:Ok:>[0])_R <Int>_N
                               :: (Ok (Syn :**: (:>)) Int :: Constraint)
                                  ~R#
                                  ((&+&) Yes1 GenBuses Int :: Constraint))))
                   `cast` ((Sym R:Ok(->)[0] <((Int, Int), Int)>_N,
                            Sym R:Ok(->)[0] <Int>_N)_R
                           :: ((Yes1 ((Int, Int), Int), Yes1 Int) :: Constraint)
                              ~R#
                              ((Ok (->) ((Int, Int), Int), Ok (->) Int) :: Constraint)))
                  ww_ww1_ww2_ww3_Bh)
        of wild_XoG8 { I# x1_Xo4b ->
        case exr
               @ (->)
               @ ((Int, Int), Int)
               @ Int
               ($p1BoolCat
                  @ (->) ($p1EqCat @ (->) @ (Either () ()) $s$fEqCat(->)a_skw8))
               (($fYes1ka @ * @ ((Int, Int), Int),
                 $p1&+&
                   @ *
                   @ Yes1
                   @ GenBuses
                   @ Int
                   (($p1(%,,%)
                       @ (Ok (Syn :**: (:>)) Int)
                       @ (Ok (Syn :**: (:>)) Int)
                       @ (Ok (Syn :**: (:>)) Bool)
                       $d(%,,%)_XrPz)
                    `cast` (Sub (D:R:Ok:**:[0] <(:>)>_N <Syn>_N)
                            ; (R:OkSyn[0] &+& D:R:Ok:>[0])_R <Int>_N
                            :: (Ok (Syn :**: (:>)) Int :: Constraint)
                               ~R#
                               ((&+&) Yes1 GenBuses Int :: Constraint))))
                `cast` ((Sym R:Ok(->)[0] <((Int, Int), Int)>_N,
                         Sym R:Ok(->)[0] <Int>_N)_R
                        :: ((Yes1 ((Int, Int), Int), Yes1 Int) :: Constraint)
                           ~R#
                           ((Ok (->) ((Int, Int), Int), Ok (->) Int) :: Constraint)))
               ww_ww1_ww2_ww3_Bh
        of wild1_Xo4g { I# y1_Xo4j ->
        let {
          $j_s4ICl :: Int# -> Int -> Int
          [LclId,
           Arity=2,
           Str=DmdType,
           Unf=Unf{Src=<vanilla>, TopLvl=False, Value=True, ConLike=True,
                   WorkFree=True, Expandable=True, Guidance=IF_ARGS [0 0] 120 0}]
          $j_s4ICl =
            \ (y1_Xo4p [OS=OneShot] :: Int#)
              (wild1_Xo4m [OS=OneShot]
                 :: Int
                 Unf=Unf{Src=InlineStable, TopLvl=False, Value=True, ConLike=True,
                         WorkFree=True, Expandable=True,
                         Guidance=ALWAYS_IF(arity=0,unsat_ok=True,boring_ok=False)
                         Tmpl= I# y1_Xo4p}) ->
              case lessThanOrEqual
                     @ (->)
                     @ Int
                     ($fOrdCat(->)a @ Int $fOrdInt)
                     (boxI x1_Xo48, boxI y1_Xo4p)
              of _ [Occ=Dead] {
                False -> wild1_Xo4m;
                True -> wild_Xo45
              } } in
        case lessThanOrEqual
               @ (->)
               @ Int
               ($fOrdCat(->)a @ Int $fOrdInt)
               (boxI x1_Xo4b, boxI y1_Xo4j)
        of _ [Occ=Dead] {
          False -> $j_s4ICl y1_Xo4j wild1_Xo4g;
          True -> $j_s4ICl x1_Xo4b wild_XoG8
        }    
        }    
        } } in
  case lessThanOrEqual
         @ (->)
         @ Int
         ($fOrdCat(->)a @ Int $fOrdInt)
         (boxI x1_insa, boxI y1_inse)
  of _ [Occ=Dead] {
    False -> $j_s4ICn y1_inse wild1_insc;
    True -> $j_s4ICn x1_insa wild_ins8
  }          
  }          
  } 
```
